### PR TITLE
Workflow: Pytest and Pylint Checks

### LINKED ISSUE #27: Setup Github Actions for Pytest
### Steps
- [x] Create a Github Action workflow for running Pytest
- [x] Configure the workflow to run on pull requests and pushes
- [x] Validate the test results to ensure code functionality

### User Story
As a developer, I need a Github Action for Pytest to automatically run tests and ensure the code is functioning correctly.

### PR DESCRIPTION
**Pull request by NabihaMoh**
Added an automated workflow for testing. It displays a badge in the README file showcasing the pytest score as percentage. closes #27

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+on: [push, pull_request]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      # Checkout the code
+      - name: Checkout
+        uses: actions/checkout@v4
+      # Install Python
```
Comment (A-Faris) [inline .github/workflows/main.yaml:19]: Nice comments, but if the comment and name is the same, is it needed?

File: .github/badges/avg_score.svg
Comment (A-Faris) [inline .github/badges/avg_score.svg:1]: Nice badges, makes it easier to view scores

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+          done
+
+      # Create .util
+      - name: Create .util
+        run: |
+          [ ! -d .util ] && mkdir -p .util || true
+
+      # Run pytest
+      - name: Run tests
+        if: '!cancelled()'
```
Comment (A-Faris) [inline .github/workflows/main.yaml:40]: What is the use of this if statement? It occurs multiple times

File: .util/parse_raw_python.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    if len(matches) == 0:
+        return [0.0]
+    return [float(match) for match in matches]
+
+
+def calculate_avg_percentage_of_passed_tests(passed_tests: int, total_tests: int) -> int:
+    '''Calculate the mean percentage passed tests given passed tests and total tests.'''
+    if passed_tests == 0:
+        return 0
+    return int(100 * passed_tests / total_tests)
```
Comment (A-Faris) [inline .util/parse_raw_python.py:66]:
```suggestion
    return 100 * passed_tests // total_tests
```

This is similar to the floor function but ignore if this isn't preferred

File: main.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+def return_4():
```
Comment (A-Faris) [inline main.py]: Make sure to add a docstring to the file and a docstring to the function even if it is for testing. The function docstring can mention how this is a function for testing

File: .github/workflows/main.yaml (hunk 1 above)
Comment (NabihaMoh) [reply .github/workflows/main.yaml:19]: Comments help provide more detail and creates separation within the file to make it easier to read.

File: .github/workflows/main.yaml (hunk 2 above)
Comment (NabihaMoh) [reply .github/workflows/main.yaml:40]: It a check that runs the step only if the run hasn't been cancelled.

File: main.py (hunk 4 above)
Comment (NabihaMoh) [reply main.py]: I did this to get a specific pylint score, so I could test the pylint acts as expected.

File: .util/parse_raw_python.py (hunk 3 above)
Comment (NabihaMoh) [reply .util/parse_raw_python.py:66]: I will keep this in mind for next time, thanks!

File: .github/badges/avg_score.svg
Comment (A-Faris) [inline .github/badges/avg_score.svg:1]: Nice badges, makes scores easier to view

File: .github/workflows/main.yaml [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        run: |
+          echo "AVG_SCORE=$(jq '.avg_score' .util/pylint_scores.json)" >> $GITHUB_ENV
+        env:
+          CODE_QUALITY: ${{ env.AVG_SCORE }}
+          BADGE_PATH: .github/badges/avg_score.svg   
+          
+      - name: Fail pylint if score below threshold 
+        if: '!cancelled()'
+        run: |
+          result=$(echo "$AVG_SCORE < 9.0" | bc -l)
```
Comment (A-Faris) [inline .github/workflows/main.yaml:119]: I'm glad this works now

File: .github/workflows/main.yaml [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+          git add .github/badges/passed_percentage.svg
+          git add .github/badges/avg_score.svg
+          git commit -m "Update scores" || true
+          echo "Event Type: ${{ github.event_name }}"
+          if [[ "${{ github.event_name }}" == "push" ]]; then
+            echo "Pushing"
+            git push
+          else
+            echo "No Push"
+          fi
```
Comment (A-Faris) [inline .github/workflows/main.yaml:187]: Good job on fixing the pushing part

File: .util/pylint_scores_test.txt
Comment (A-Faris) [inline .util/pylint_scores_test.txt:1]: I think this file needs to be deleted

Comment (NabihaMoh) [reply .util/pylint_scores_test.txt:1]: Done.

**Review by A-Faris — APPROVED** [review]
Looks good now

**Pull request merged**